packages/promotion/src/repositories/campaign.ts

```diff
@@ -8,4 +10,137 @@ export class CampaignRepository extends DALUtils.mikroOrmBaseRepositoryFactory<
     create: CreateCampaignDTO
     update: UpdateCampaignDTO
   }
->(Campaign) {}
+>(Campaign) {
+  async create(
+    data: CreateCampaignDTO[],
+    context: Context = {}
+  ): Promise<Campaign[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+    const promotionIds: string[] = []
+    const campaignIdentifierPromotionsMap = new Map<string, string[]>()
+
+    data.forEach((campaignData) => {
+      const campaignPromotionIds =
+        campaignData.promotions?.map((p) => p.id) || []
+
+      promotionIds.push(...campaignPromotionIds)
+
+      campaignIdentifierPromotionsMap.set(
+        campaignData.campaign_identifier,
+        promotionIds
+      )
+
+      delete campaignData.promotions
+    })
+
+    const existingPromotions = await manager.find(Promotion, {
+      id: promotionIds,
+    })
+
+    const existingPromotionsMap = new Map<string, Promotion>(
+      existingPromotions.map((promotion) => [promotion.id, promotion])
+    )
+
+    const createdCampaigns = await super.create(data, context)
+
+    for (const createdCampaign of createdCampaigns) {
+      const campaignPromotionIds =
+        campaignIdentifierPromotionsMap.get(
+          createdCampaign.campaign_identifier
+        ) || []
+
+      if (!campaignPromotionIds.length) {
+        continue
+      }
+
+      for (const campaignPromotionId of campaignPromotionIds) {
+        const promotion = existingPromotionsMap.get(campaignPromotionId)
+
+        if (!promotion) {
+          continue
+        }
+
+        createdCampaign.promotions.add(promotion)
+      }
+    }
+
+    return createdCampaigns
+  }
+
+  async update(
+    data: UpdateCampaignDTO[],
+    context: Context = {}
+  ): Promise<Campaign[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+    const promotionIds: string[] = []
+    const campaignIds: string[] = []
+    const campaignPromotionIdsMap = new Map<string, string[]>()
+
+    data.forEach((campaignData) => {
+      const campaignPromotionIds =
+        campaignData.promotions?.map((p) => p.id) || []
+
+      campaignIds.push(campaignData.id)
+      promotionIds.push(...campaignPromotionIds)
+      campaignPromotionIdsMap.set(campaignData.id, campaignPromotionIds)
+
+      delete campaignData.promotions
+    })
+
+    const existingCampaigns = await manager.find(
+      Campaign,
+      { id: campaignIds },
+      { populate: ["promotions"] }
+    )
+
+    const existingPromotions = await manager.find(Promotion, {
+      id: promotionIds,
+    })
+
+    const existingCampaignsMap = new Map<string, Campaign>(
+      existingCampaigns.map((campaign) => [campaign.id, campaign])
+    )
+
+    const existingPromotionsMap = new Map<string, Promotion>(
+      existingPromotions.map((promotion) => [promotion.id, promotion])
+    )
+
+    const updatedCampaigns = await super.update(data, context)
+
+    for (const updatedCampaign of updatedCampaigns) {
+      const promotionIdsToAdd =
```

Comment:
suggestion: the naming here is a bit confusing, since the array seems to contain both existing promotion ids to keep and new ids to add.

```suggestion
      const upsertPromotionIds =
```